aten/src/ATen/core/TransformationHelper.h

```diff
@@ -68,6 +66,46 @@ C10_HOST_DEVICE inline T uniform_int(V val) {
   }
 }
 
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <float, uint32_t>
+ */
+template <>
+C10_HOST_DEVICE inline float uniform_int(uint32_t val) {
+  return static_cast<float>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<float>::digits) + 1));
+}
+
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <float, uint64_t>
+ */
+template <>
+C10_HOST_DEVICE inline float uniform_int(uint64_t val) {
+  return static_cast<float>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<float>::digits) + 1));
+}
+
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <double, uint32_t>
+ */
+template <>
+C10_HOST_DEVICE inline double uniform_int(uint32_t val) {
+  return static_cast<double>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<double>::digits) + 1));
+}
+
+/**
+ * An overloaded transformation function for `torch.Tensor.random_()`, when used without specifying `from` and `to`,
+ * added to fix compiler warnings reported in GitHub issue 46391.
+ * Here, <T, V> is <double, uint64_t>
```

Comment:
Small nits:
 - Can you fold 4 comment blocks into 4?
 - This implies that those functions are only called for unsigned integer type. Is it possible to add a compile time check that those templates could not be instantiated with signed integral type?
 P.S. Also, please explain why something like the following wouldn't work (instead of 4 explicit full specializations):
```
template<typename T, typename V>
C10_HOST_DEVICE inline std::enable_if<std::is_floating_point<T>::value, T>::type uniform_int(V val) {
    return static_cast<T>(val % static_cast<uint64_t>((1ULL << std::numeric_limits<T>::digits) + 1));
}

Reply:
Thank you for your prompt review, and for this learning opportunity!

I had never used enable_if before, so I agree that your suggestion is way better than using full specializations.

As for the compile-time check for disallowing instantiation with a signed integral type, could you please suggest what should the behavior be in that case? Please advise if an assertion would be appropriate. Thanks!